nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ScriptedTransformRecord.java

```diff
@@ -0,0 +1,398 @@
+                                        + "at least one of which was not a Record but instead was: " + returnValue);
+                                }
+
+                                writer.write((Record) element);
+                            }
+
+                            continue;
+                        }
+
+                        // Ensure that the value returned from the script is either null or a Record
+                        throw new RuntimeException("Evaluated script against Record number " + index + " of " + flowFile
+                            + " but instead of returning a Record, script returned a value of: " + returnValue);
+                    }
+
+                    // Add WriteResults to the attributes to be added to the FlowFile
+                    final WriteResult writeResult = writer.finishRecordSet();
+                    attributesToAdd.put("mime.type", writer.getMimeType());
+                    attributesToAdd.putAll(writeResult.getAttributes());
+                    attributesToAdd.put("record.count", String.valueOf(writeResult.getRecordCount()));
+                } catch (final MalformedRecordException | SchemaNotFoundException | ScriptException e) {
+                    throw new ProcessException(e);
+                }
+            });
+
+            // Add the necessary attributes to the FlowFile and transfer to success
+            session.putAllAttributes(flowFile, attributesToAdd);
+            session.transfer(flowFile, REL_SUCCESS);
+
+            final long transformCount = recordIndex.get() - dropCount.get();
+            getLogger().info("Successfully transformed {} Records and dropped {} Records for {}", new Object[] {transformCount, dropCount.get(), flowFile});
+            session.adjustCounter("Records Transformed", transformCount, true);
+            session.adjustCounter("Records Dropped", dropCount.get(), true);
+
+            final long millis = System.currentTimeMillis() - startMillis;
+            session.getProvenanceReporter().modifyContent(flowFile, "Transformed " + transformCount + " Records, Dropped " + dropCount.get() + " Records", millis);
+        } catch (final ProcessException e) {
+            getLogger().error("After processing {} Records, encountered failure when attempting to transform {}", new Object[] {recordIndex.get(), flowFile}, e.getCause());
+            session.transfer(flowFile, REL_FAILURE);
+        } catch (final Exception e) {
+            getLogger().error("After processing {} Records, encountered failure when attempting to transform {}", new Object[] {recordIndex.get(), flowFile}, e);
+            session.transfer(flowFile, REL_FAILURE);
+        }
+    }
+
+    private ScriptEvaluator createEvaluator(final ScriptEngine scriptEngine, final FlowFile flowFile) throws ScriptException {
+        if (PYTHON_SCRIPT_LANGUAGE.equalsIgnoreCase(scriptEngine.getFactory().getLanguageName())) {
```

Comment:
@mattyb149 we deprecated support for Jython, didn't we, in `ExecuteScript`?

Reply:
My memory on the matter is vague, but I remember there being some serious bug in Jython that made us have to exclude Jython.

Reply:
We don't really have a way of deprecating such a thing. We generally recommend people use Groovy instead of Jython because the performance difference is pretty huge. And Jython definitely has some key limitations, such as the inability to call "C Python" libraries. But I don't recall any bugs that are critical enough for us to say that it shouldn't be used. Unless there's something I missed. @mattyb149 does that ring a bell for you?

Reply:
The Jython issue for scripted components was mostly because there's a bug where we can't subclass interfaces that have default methods. For the scripted components whose only job pretty much is to implement such interfaces (RecordReaderFactory, etc.) Jython doesn't make much sense. But Python/Jython shows up a lot in our community, so for this particular processor I think it's a good idea to support Jython here. I imagine quite a few folks will find this very helpful, to quickly script up a transformation in a language they're comfortable with, without having to worry about implementing much of the NiFi API.

Reply:
Thanks for clarifying @mattyb149!